core/trino-main/src/main/java/io/trino/operator/DirectExchangeClient.java

```diff
@@ -389,6 +398,7 @@ private synchronized void clientFinished(HttpPageBufferClient client)
     {
         requireNonNull(client, "client is null");
         if (completedClients.add(client)) {
+            runningClients.remove(client);
```

Comment:
nit: you could add check state that client was actually removed. Similar to adding clients to `queuedClients`